### PR TITLE
pkg/nodediscovery,daemon: modularize node discovery

### PR DESCRIPTION
**Pull request by dylandreimerink**
This commit modularizes the node discovery package. Before node discovery was created by the daemon, but since all parameters needed are already in hive we can create the node discovery in the hive to.

We also split off the creation of local node config into its own cell since there are a few components such as the loader that are interested in the local node config without needing the full node discovery.

```release-note
modularize node discovery
```

### CODE REVIEW

File: daemon/cmd/cells.go [hunk 1]
```diff
@@ -244,6 +245,9 @@ var (
 
 		// The device reloader reloads the datapath when the devices change at runtime.
 		cell.Invoke(registerDeviceReloader),
+
+		// The node discovery cell provides the local node configuration and node discovery.
```
Comment (joamaki) [inline daemon/cmd/cells.go]: Perhaps could spell out what we mean by "node discovery" here, e.g. it propagates changes in the local node towards other Cilium nodes via CiliumNode CRD and kvstore.

File: pkg/nodediscovery/cell.go [hunk 2]
```diff
@@ -0,0 +1,14 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package nodediscovery
+
+import "github.com/cilium/cilium/pkg/hive/cell"
+
+var Cell = cell.Module(
```
Comment (joamaki) [inline pkg/nodediscovery/cell.go:10]: Let's document the different things NodeDiscovery does with a comment here. E.g. an expanded version of what we have in daemon/cmd/cells.go.

File: pkg/nodediscovery/cell.go [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+	"nodediscovery",
+	"Node Discovery",
```
Comment (joamaki) [inline pkg/nodediscovery/cell.go]: No need to repeat the module id in the title ("nodediscovery: Node Discovery"). Instead the title should be something like "propagates local node information to other nodes" or some such.

File: pkg/nodediscovery/localnodeconfig.go [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+	"github.com/cilium/cilium/pkg/option"
+)
+
+func NewLocalNodeConfig(mtu mtu.MTU, config *option.DaemonConfig) datapath.LocalNodeConfiguration {
+	auxPrefixes := []*cidr.CIDR{}
+
+	if config.IPv4ServiceRange != AutoCIDR {
+		serviceCIDR, err := cidr.ParseCIDR(config.IPv4ServiceRange)
+		if err != nil {
+			log.WithError(err).WithField(logfields.V4Prefix, config.IPv4ServiceRange).Fatal("Invalid IPv4 service prefix")
```
Comment (joamaki) [inline pkg/nodediscovery/localnodeconfig.go]: Constructors should not fatal but should instead return an error.

File: pkg/nodediscovery/localnodeconfig.go [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+		}
+
+		auxPrefixes = append(auxPrefixes, serviceCIDR)
+	}
+
+	if config.IPv6ServiceRange != AutoCIDR {
+		serviceCIDR, err := cidr.ParseCIDR(config.IPv6ServiceRange)
+		if err != nil {
+			log.WithError(err).WithField(logfields.V6Prefix, config.IPv6ServiceRange).Fatal("Invalid IPv6 service prefix")
```
Comment (joamaki) [inline pkg/nodediscovery/localnodeconfig.go]: Ditto here

**Review by joamaki — CHANGES_REQUESTED** [review]

File: pkg/nodediscovery/cell.go (hunk 3 above)
Comment (dylandreimerink) [reply pkg/nodediscovery/cell.go]: We should really rename the second argument to "description" or "summary", to me a title is the name of the thing, not a description of what it does

Comment (joamaki) [reply pkg/nodediscovery/cell.go]: oh yeah, you're totally right. I'll make this change in `cilium/hive` and expand the `cell.Module` to show an example.

File: pkg/nodediscovery/cell.go (hunk 2 above)
Comment (joamaki) [reply pkg/nodediscovery/cell.go:10]: I meant documenting `var Cell`, e.g. like in `pkg/endpointmanager/cell.go`. Use-case is when hovering over `nodediscovery.Cell` in `daemon/cmd/cells.go` we'd get a bit more detailed description of what objects are provided from here and what they do. Sorry for the nit-picking. Trying to make the habit stick :)

File: pkg/nodediscovery/cell.go [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+
+import "github.com/cilium/cilium/pkg/hive/cell"
+
+var Cell = cell.Module(
+	"nodediscovery",
+	"Communicate changes in local node information to the API server or KVStore",
+
+	// Node discovery communicates changes in local node information to the API server or KVStore
+	cell.Provide(NewNodeDiscovery),
+	// LocalNodeConfig provides a subset of the deamon config with a little pre-processing
```
Comment (joamaki) [inline pkg/nodediscovery/cell.go]: s/deamon config/DaemonConfig/

**Review by joamaki — APPROVED** [review]

File: pkg/nodediscovery/cell.go (hunk 2 above)
Comment (dylandreimerink) [reply pkg/nodediscovery/cell.go:10]: Right, I had misunderstood what you meant by "what we have in daemon/cmd/cells.go"

**Pull request merged**